presto-spark-base/src/main/java/com/facebook/presto/spark/planner/PrestoSparkRddFactory.java

```diff
@@ -379,14 +383,85 @@ private PrestoSparkTaskSourceRdd createTaskSourcesRdd(
         return assignPartitionedSplits(session, partitioning, splits);
     }
 
-    private static SetMultimap<Integer, ScheduledSplit> assignSourceDistributionSplits(Session session, List<ScheduledSplit> splits)
+    @VisibleForTesting
+    public static SetMultimap<Integer, ScheduledSplit> assignSourceDistributionSplits(Session session, List<ScheduledSplit> splits)
     {
-        int taskCount = getSparkInitialPartitionCount(session);
-        checkArgument(taskCount > 0, "taskCount must be greater then zero: %s", taskCount);
         ImmutableSetMultimap.Builder<Integer, ScheduledSplit> result = ImmutableSetMultimap.builder();
-        for (int splitIndex = 0; splitIndex < splits.size(); splitIndex++) {
-            result.put(splitIndex % taskCount, splits.get(splitIndex));
+
+        long maxSplitsSizeInBytesPerPartition = getMaxSplitsDataSizePerSparkPartition(session).toBytes();
+        checkArgument(maxSplitsSizeInBytesPerPartition > 0,
+                "maxSplitsSizeInBytesPerPartition must be greater than zero: %s", maxSplitsSizeInBytesPerPartition);
+        int initialPartitionCount = getSparkInitialPartitionCount(session);
+        checkArgument(initialPartitionCount > 0,
+                "initialPartitionCount must be greater then zero: %s", initialPartitionCount);
+
+        boolean splitsDataSizeAvailable = splits.stream()
+                .allMatch(split -> split.getSplit().getConnectorSplit().getSplitSizeInBytes().isPresent());
+        if (!splitsDataSizeAvailable) {
+            for (int splitIndex = 0; splitIndex < splits.size(); splitIndex++) {
+                result.put(splitIndex % initialPartitionCount, splits.get(splitIndex));
+            }
+        }
+        else {
```

Comment:
nit: Let's simply `return result.build();` instead of adding one more level of indentation